neps/nep-0488.md

```diff
@@ -0,0 +1,1222 @@
+The encoding for $x$ and $y$ follows the rules detailed in the "Extension Fields Elements $F_{p^2}$" section.
+
+*The second-highest bit* within the encoding serves to signify a point at infinity. 
+When this bit is set to 1, it designates an infinity point. 
+In this case, all other bits should be set to 0.
+
+Encoding the point at infinity:
+
+```bash
+let x: [u8; 192] = [0; 192];
+x[0] = x[0] | 0x40;
+```
+
+#### Compressed points on twisted curve $E'(F_{p^2})$
+
+The points on the curve are represented by affine coordinates: $(x: F_{p^2}, y: F_{p^2})$. 
+Elements from $E'(F_{p^2})$ in compressed form are encoded as [u8; 96], 
+with big-endian encoded $x \in F_{p^2}$. 
+The $y$ coordinate is determined using the formula: $y = \pm \sqrt{x^3 + 4(u + 1)}$.
+
+- The highest bit indicates if the point is encoded in compressed form and should be set to 1.
+- The second-highest bit marks the point at infinity (if set to 1).
+  - For the point at infinity, all bits except the first two should be set to 0; other encodings should be considered as incorrect.
+- To represent the sign of $y$, the third-highest bit in the x encoding is utilized.
+  - If the bit is 0, $y$ is positive; if 1, $y$ is negative. We'll consider the number positive by taking the smallest value between $y$ and $-y$: first compare $c_1$, then $c_0$.
+
+The encoding of $x \in F_{p^2}$ as [u8; 96] bytes follows the rules from the section “Extension Fields Elements $F_{p^2}$”.
+
+Encoding a point on $E'(F_{p^2})$ with a negative $y$ coordinate:
+
+```rust
+let x: [u8; 96] = encodeFp2(x);
+x[0] = x[0] | 0x80;
+x[0] = x[0] | 0x20;
+```
+
+Encoding the point at infinity:
+
+```rust
+let x: [u8; 96] = [0; 96];
+x[0] = x[0] | 0x80;
+x[0] = x[0] | 0x40;
+```
+
+#### ERROR_CODE
+
+Validating the input for the host functions within the contract can consume significant gas.
+For instance, verifying if a point belongs to the subgroup is gas-consuming. 
+If an error is returned by the near host function, the entire execution is reverted. 
+To mitigate this, when the input verification is complex, the host function 
+will successfully complete its work but return an ERROR_CODE. 
+This enables users to handle error cases independently. It's important to note that host functions
+might terminate with an error if it's straightforward to avoid it (e.g., incorrect input size).
+
+The ERROR_CODE is an u64 and can hold the following values:
+
+- 0: No error, execution was successful. For `bls12381_pairing_check` function, the pairing result equals the multiplicative identity.
+- 1: Execution finished with error due to:
+  - Incorrect encoding (e.g., incorrectly set compression/decompression bit, coordinate >= p, etc.). 
+  - A point not on the curve (where applicable). 
+  - A point not in the expected subgroup (where applicable).
+- 2: Can be returned only in `bls12381_pairing_check`. No error, execution was successful, but the pairing result doesn't equal the multiplicative identity. 
```

Comment:
I don't understand this comment. Why we should be consistent with Linux errno? The linux errno can be negative? We can not use the negative numbers because we're returning the unsigned type.

Reply:
Linux typically uses functions that return a positive number (often 1, sometimes other values) for success, and a negative number for failure, where which exact negative number defines the details of the error.

This is good because it extends seamlessly to cases where we also need to return either an error or a number between [0, i64::MAX].

So I’d suggest returning `i32` instead of `u32`, and renumbering the return values to be -1 for error, 0 for successful execution; and, for bls12381_pairing_check, 1 for "the result didn’t match".

Incidentally, `-1 as u32` is 0xFFFFFFFF, which means we could easily change the spec later on, if we were to need a function that can error with return values that go up to 0xFFFFFFFE.

Reply:
In nearcore only two returned types are supported: u64 and u32. It is not possible to return the signed type. We can indeed represent the negative numbers in unsigned types, as you showed, but I am afraid that it can make the logic even more complicated. 

Your suggestion makes a lot of sense, and it is logical when we use the negative numbers for error cases and the positive ones for the results. But in this specific case, I prefer not to change anything to avoid casting between signed and unsigned types.

Reply:
> In nearcore only two returned types are supported: u64 and u32.

Near-vm (or basically any spec-compliant wasm interpreter) should work without any issue with `i32` parameters, so long as the host function is properly defined with the right types on both sides. For near-vm (or any wasm interpreter, as per the wasm spec) the parameter is just "a 32-bit value", and how to interpret it is just convention on both sides, with no preference toward one or the other.

Now, maybe I missed something, but I’d be very surprised if we were unable to have host functions that take/return `i32`

(That said, I totally understand your reluctance now! If we’re indeed unable to return `i32` then the added complexity would not make sense for now)